lib/connection/tcp.ex

```diff
@@ -10,37 +10,42 @@ defmodule LogstashJson.TCP.Connection do
   size is reached after which incoming messages are dropped.
   """
 
-  def start_link(host, port, opts, timeout \\ 1_000, buffer_max \\ 10_000) do
-    Connection.start_link(__MODULE__, {host, port, opts, timeout, buffer_max})
+  @connection_opts [active: false, mode: :binary, keepalive: true, packet: 0]
+
+  def start_link(host, port, queue, timeout \\ 1_000) do
+    Connection.start_link(__MODULE__, {host, port, queue, timeout})
   end
 
-  @doc "Asynchronous sending of a message"
-  def send(conn, data), do: Connection.cast(conn, {:send, data})
+  @doc "Send message to logstash backend"
+  def send(conn, data, timeout \\ 5_000) do
+    Connection.call(conn, {:send, data}, timeout)
+  end
 
   @doc "Close connection"
   def close(conn), do: Connection.call(conn, :close)
 
   @doc "Update configuration and reconnect"
-  def configure(conn, host, port, opts) do
-    Connection.call(conn, {:configure, host, port, opts})
+  def configure(conn, host, port) do
+    Connection.call(conn, {:configure, host, port})
   end
 
-  def init({host, port, opts, timeout, buffer_max}) do
+  def init({host, port, queue, timeout}) do
+    LogstashJson.TCP.Connection.Worker.start_link(self(), queue)
+
     state = %{
       host: host,
       port: port,
-      opts: opts,
       timeout: timeout,
-      sock: nil,
-      buffer: [],
-      buffer_max: buffer_max}
+      sock: nil}
     {:connect, :init, state}
   end
 
-  def connect(_, %{sock: nil, host: host, port: port, opts: opts, timeout: timeout} = state) do
-    case :gen_tcp.connect(host, port, [active: false] ++ opts, timeout) do
+  def connect(_, %{sock: nil, host: host, port: port, timeout: timeout} = state) do
+    case :gen_tcp.connect(host, port, @connection_opts, timeout) do
       {:ok, sock} ->
         {:ok, %{state | sock: sock}}
+      {:error, :econnrefused} ->
+        {:backoff, 1000, state}
```

Comment:
@tfardig Updated the error logging, it now logs once when a failed connection attempt happens, and then once again per message, and we don't treat :econnrefused specially anymore. Logs now look like this when started on a machine with no logstash connection:

```
iex(1)> Elixir.LogstashJson.TCP.Connection[2]: localhost:4560 connection failed: connection refused
Elixir.LogstashJson.TCP.Connection[1]: localhost:4560 connection failed: connection refused

iex(3)> require Logger
Logger
iex(4)> Logger.info "hi"
:ok
Elixir.LogstashJson.TCP.Connection[2]: localhost:4560 connection failed: connection refused
iex(5)> Logger.info "hi"
:ok
Elixir.LogstashJson.TCP.Connection[2]: localhost:4560 connection failed: connection refused
iex(6)> 
```

Do you think it looks good?